core-data/reducer.js

```diff
@@ -129,23 +129,85 @@ function entity( entityConfig ) {
 	};
 }
 
-const entitiesByKind = groupBy( entitiesConfig, 'kind' );
-export const entities = combineReducers( Object.entries( entitiesByKind ).reduce( ( memo, [ kind, subEntities ] ) => {
-	const kindReducer = combineReducers( subEntities.reduce(
-		( kindMemo, entityConfig ) => ( {
-			...kindMemo,
-			[ entityConfig.name ]: entity( entityConfig ),
-		} ),
-		{}
-	) );
+/**
+ * Reducer keeping track of the registered entities.
+ *
+ * @param {Object} state  Current state.
+ * @param {Object} action Dispatched action.
+ *
+ * @return {Object} Updated state.
+ */
+export function entitiesConfig( state = defaultEntities, action ) {
+	switch ( action.type ) {
+		case 'ADD_ENTITIES':
+			return [
+				...state,
+				...action.entities,
+			];
+	}
 
-	memo[ kind ] = kindReducer;
-	return memo;
-}, {} ) );
+	return state;
+}
+
+/**
+ * Reducer keeping track of the registered entities config and data.
+ *
+ * @param {Object} state  Current state.
+ * @param {Object} action Dispatched action.
+ *
+ * @return {Object} Updated state.
+ */
+export const entities = ( state = {}, action ) => {
+	const newConfig = entitiesConfig( state.config, action );
+
+	// Generates a dynamic reducer for the entities
+	const entitiesByKind = groupBy( newConfig, 'kind' );
```

Comment:
Does this need to be run on every dispatch? Or could we limit it to those which would actually impact the configuration to justify dynamically creating new reducers?

Reply:
Do you I should use a local variable to keep track of the current reducer?

Reply:
Maybe it's included as part of the `state` value?

Reply:
It does seem strange, though accurate if we're saying that the reducer itself is dependent on state, that it is itself state.

Or, if it's a derivation, then a selector may be appropriate? 

Entering unfamiliar territory for me with this dynamic reducer 😄